File: algorithms/15_ruby/strings/manacher.rb
def manacher(s)
  t = "#" + s.chars.join("#") + "#"
  n = t.length
  p = Array.new(n, 0)
  c = r = 0
  (0...n).each do |i|
    p[i] = [r - i, p[2 * c - i]].min if i < r
    while i + p[i] + 1 < n && i - p[i] - 1 >= 0 &&
          t[i + p[i] + 1] == t[i - p[i] - 1]
      p[i] += 1
    end
    if i + p[i] > r
      c = i
      r = i + p[i]
    end
  end
  p.max
end

raise "manacher wrong" unless manacher("abba") == 4
raise "manacher odd wrong" unless manacher("racecar") == 7
puts "[Ruby Manacher] Longest palindromic substring verified"

File: algorithms/23_elixir/strings/manacher.ex
defmodule Manacher do
  def longest_palindrome(s) do
    t = "#" <> (s |> String.graphemes() |> Enum.join("#")) <> "#"
    chars = String.graphemes(t)
    n = length(chars)
    {p, _} =
      Enum.reduce(0..(n - 1), {List.duplicate(0, n), {0, 0}}, fn i, {p, {c, r}} ->
        pi = if i < r, do: min(r - i, Enum.at(p, 2 * c - i)), else: 0
        pi = extend(chars, i, pi)
        p = List.replace_at(p, i, pi)
        {c, r} = if i + pi > r, do: {i, i + pi}, else: {c, r}
        {p, {c, r}}
      end)
    Enum.max(p)
  end

  defp extend(chars, i, pi) do
    n = length(chars)
    cond do
      i + pi + 1 < n and i - pi - 1 >= 0 and
          Enum.at(chars, i + pi + 1) == Enum.at(chars, i - pi - 1) ->
        extend(chars, i, pi + 1)
      true -> pi
    end
  end
end

unless Manacher.longest_palindrome("abba") == 4, do: raise("manacher even wrong")
unless Manacher.longest_palindrome("racecar") == 7, do: raise("manacher odd wrong")
IO.puts("[Elixir Manacher] Longest palindromic substring verified")
